### PR TITLE
Compile rustdoc less often.

### LINKED ISSUE #70799: x.py: running "doc" after "build" always rebuilds rustdoc
But just re-running `./x.py doc` repeatedly works as expected... nothing gets rebuilt.

### PR DESCRIPTION
**Pull request by ehuss**
Previously rustdoc was built 3 times with `x.py test`:

1. stage2 (using stage1 compiler) for compiletest tests (stage1-tools copied to stage2).
2. stage1 (using stage0 compiler) for std crate tests (stage0-tools copied to stage1).
3. stage2 test (using stage2 compiler) for rustdoc crate tests and error_index_generator (stage2-tools).

This PR removes the majority of number 3, where it will instead use the stage1 compiler, which will share the artifacts from number 1.

This matches the behavior of the libstd crate tests. I don't think it is entirely necessary to run the tests using stage2.

At `-j2`, the last build step goes from about 300s to 70s on my machine. It's not a huge win, but shaving 4 minutes isn't bad.

The other two builds would be pretty difficult (or undesired or impossible) to unify. It looks like std tests use stage1 very intentionally (see `force_use_stage1` and its history), and compiletests use the top stage very intentionally.

Unfortunately the linkchecker builds all docs at stage2 (stage2-tools), which means a few build script artifacts are not shared. It's not really clear to me how to fix that (because it uses `default_doc`, there doesn't seem to be any control over the stages).

---

For `x.py doc`, rustdoc was previously built three times (with compiler-docs):

1. stage2 (using stage1 compiler) for normal documentation output (stage1-tools copied to stage2).
2. stage1 (using stage0 compiler) for compiler-docs
3. stage2 (using stage2 compiler) for error_index_generator (stage2-tools)

This PR combines these so that they consistently use the "top stage" rustdoc. I don't know why the compiler-docs was written to use stage minus one, but it seems better to be consistent across the doc steps.

---

I've tried to test this with a variety of commands (`x.py doc`, `x.py test`, different `--stage` flags, `full-bootstrap`, setting `--target`, etc.) to try to make sure there aren't significant regressions here. It's tricky since there are so many variables, and this stuff is difficult for me to fully understand.

Closes #70799 (I think)

### CODE REVIEW

File: src/bootstrap/test.rs [hunk 1]
```diff
@@ -1825,6 +1833,32 @@ impl Step for CrateRustdoc {
...
+        // librustdoc. There is only one test, and it doesn't look too
+        // important. There might be other ways to avoid this, but it seems
+        // pretty convoluted.
+        //
+        // See also https://github.com/rust-lang/rust/issues/13983 where the
+        // host vs target dylibs for rustdoc are consistently tricky to deal
+        // with.
+        let mut dylib_path = dylib_path();
+        dylib_path.insert(0, PathBuf::from(&*builder.sysroot_libdir(compiler, target)));
+        cargo.env(dylib_path_var(), env::join_paths(&dylib_path).unwrap());
```
Comment (Mark-Simulacrum) [inline src/bootstrap/test.rs:1860]: I would be on board with deleting that test - perhaps in a separate PR - and adding a tidy lint or something banning doc tests in librustdoc.

**Review by Mark-Simulacrum — COMMENTED** [review]
Can you check that we still avoid building rustdoc in stage0? (i.e. `x.py test --stage 0 src/libstd`)

Other than that though this looks reasonable, though I agree it's pretty annoying to try and get this right. Maybe we can eventually come up with some kind of solution to that.

**Comment by ehuss** [comment]
> Can you check that we still avoid building rustdoc in stage0? (i.e. `x.py test --stage 0 src/libstd`)

I can confirm that it uses the beta rustc/rustdoc.

As for testing, should I add a few, similar to the builder tests? I'm not sure if it will really capture everything (like the implicit rustdoc dependencies in error_index_generator), but I'd be happy to add them.

**Comment by Mark-Simulacrum** [comment]
Yeah, that'd be great! You should be able to pretty much add to the existing tests fairly easily, and it should capture ~everything I believe in terms of dependencies.

**Comment by ehuss** [comment]
OK, I added a few tests.  I feel like there could be a lot more, but this seemed like a good start.

**Comment by Mark-Simulacrum** [comment]
Looks like a great start, yes. Unfortunate about the -1 behavior on rustdoc::run but not sure we can do much... well, for a future PR anyway. Thanks! @bors r+

**Pull request merged**